google/resource_compute_project_metadata_item.go

```diff
@@ -0,0 +1,189 @@
+			"project": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"key": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"value": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceComputeProjectMetadataItemCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	projectID, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	key := d.Get("key").(string)
+	val := d.Get("value").(string)
+
+	err = updateComputeCommonInstanceMetadata(config, projectID, key, nil, &val)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(key)
+
+	return nil
+}
+
+func resourceComputeProjectMetadataItemRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	projectID, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Loading project metadata: %s", projectID)
+	project, err := config.clientCompute.Projects.Get(projectID).Do()
+	if err != nil {
+		return fmt.Errorf("Error loading project '%s': %s", projectID, err)
+	}
+
+	md := flattenComputeMetadata(project.CommonInstanceMetadata.Items)
+	val, ok := md[d.Id()]
+	if !ok {
+		// Resource no longer exists
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("key", d.Id())
+	d.Set("value", val)
+
+	return nil
+}
+
+func resourceComputeProjectMetadataItemUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	projectID, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	if d.HasChange("value") {
+		key := d.Get("key").(string)
+		p, n := d.GetChange("value")
+		prev := p.(string)
+		new := n.(string)
+
+		err = updateComputeCommonInstanceMetadata(config, projectID, key, &prev, &new)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func resourceComputeProjectMetadataItemDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	projectID, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	key := d.Get("key").(string)
+	val := d.Get("value").(string)
+
+	err = updateComputeCommonInstanceMetadata(config, projectID, key, &val, nil)
+	if err != nil {
+		return err
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func updateComputeCommonInstanceMetadata(config *Config, projectID string, key string, beforeVal *string, afterVal *string) error {
+	updateMD := func() error {
+		log.Printf("[DEBUG] Loading project metadata: %s", projectID)
+		project, err := config.clientCompute.Projects.Get(projectID).Do()
+		if err != nil {
+			return fmt.Errorf("Error loading project '%s': %s", projectID, err)
+		}
+
+		// Should the ID contain the project name?
+		md := flattenComputeMetadata(project.CommonInstanceMetadata.Items)
+
+		val, ok := md[key]
+
+		if !ok {
+			if afterVal == nil {
+				// Asked to set no value and we didn't find one - we're done
+				return nil
+			}
+			if beforeVal != nil {
```

Comment:
When I wrote this, it wasn't clear to me when a refresh occurred. Does it _always_ occur before a write?

If so, then the race condition you listed above is the only one I can think of. If that's not worth checking then I'm not sure it's worth enforcing a `beforeVal` at all, right?